applications/precompiled_pva/cpp/pva_unsharp_mask/c_model/unsharp_mask_ref.cpp

```diff
@@ -0,0 +1,146 @@
+ * limitations under the License.
+ */
+
+#include "unsharp_mask_ref.hpp"
+
+namespace model {
+
+constexpr int16_t UnsharpMaskRef::m_unsharpMaskFilter[];
+
+void UnsharpMaskRef::process(uint8_t *src, uint8_t *dst)
+{
+    rgb888_to_nv24(src, m_filterInLumaPlanePtr, m_chromaPlanePtr, m_width, m_height, m_inputLinePitch, m_width);
+    filter2d(m_filterInLumaPlanePtr, m_filterOutLumaPlanePtr, m_kernelSize, m_unsharpMaskFilter, m_quantizationBits, m_width, m_height, m_width, m_width);
+    nv24_to_rgb888(m_filterOutLumaPlanePtr, m_chromaPlanePtr, dst, m_width, m_height, m_width, m_outputLinePitch);
+}
+
+/**
+ *  Y = ( (  66 * R + 129 * G +  25 * B + 128) >> 8) + 16
+ * Cb = ( ( -38 * R -  74 * G + 112 * B + 128) >> 8) + 128
+ * Cr = ( ( 112 * R -  94 * G -  18 * B + 128) >> 8) + 128
+ */
+void UnsharpMaskRef::rgb888_to_nv24(uint8_t *rgbSrc, uint8_t *lumaDst, uint8_t *chromaDst, int32_t width, int32_t height, int32_t srcLinePitch, int32_t dstLinePitch)
+{
+    for (int32_t y = 0; y < height; y++)
+    {
+        for (int32_t x = 0; x < width; x++)
+        {
+            int32_t srcIdx = y * srcLinePitch + x;
+            int32_t dstIdx = y * dstLinePitch + x;
+
+            int16_t r = rgbSrc[3 * srcIdx];
+            int16_t g = rgbSrc[3 * srcIdx + 1];
+            int16_t b = rgbSrc[3 * srcIdx + 2];
+
+            // Y
+            int32_t y = 66 * r + 129 * g + 25 * b + 128;
+            y         = (y >> 8) + 16;
+            // Cb
+            int32_t cb = -38 * r - 74 * g + 112 * b + 128;
+            cb         = (cb >> 8) + 128;
+            // Cr
+            int32_t cr = 112 * r - 94 * g - 18 * b + 128;
+            cr         = (cr >> 8) + 128;
+
+            lumaDst[dstIdx]           = y;
+            chromaDst[2 * dstIdx]     = cb;
+            chromaDst[2 * dstIdx + 1] = cr;
+        }
+    }
+}
+
+int32_t clamp(int32_t a)
+{
+    const int32_t t = ((a <= 0) ? 0 : a);
+    return t >= 255 ? 255 : t;
+}
+
+/**
+ * C = Y  -  16
+ * D = Cb - 128
+ * E = Cr - 128
+ *
+ * R = clamp(( 74 * C   +    0 * D  +  102 * E  +  32) >> 6)
+ * G = clamp(( 74 * C   -   25 * D  -   52 * E  +  32) >> 6)
+ * B = clamp(( 74 * C   +  129 * D  +    0 * E  +  32) >> 6)
+ *
+ * clamp() denotes clipping a value to the range of 0 to 255
+ */
+void UnsharpMaskRef::nv24_to_rgb888(uint8_t *lumaSrc, uint8_t *chromaSrc, uint8_t *rgbDst, int32_t width, int32_t height, int32_t srcLinePitch, int32_t dstLinePitch)
+{
+    for (int32_t y = 0; y < height; y++)
+    {
+        for (int32_t x = 0; x < width; x++)
+        {
+            int32_t srcIdx = y * srcLinePitch + x;
+            int32_t dstIdx = y * dstLinePitch + x;
+
+            int16_t c = lumaSrc[srcIdx] - 16;
+            int16_t d = chromaSrc[2 * srcIdx] - 128;
+            int16_t e = chromaSrc[2 * srcIdx + 1] - 128;
+
+            // R
+            int32_t p0 = 74 * c + 32;
+            int32_t p1 = (p0 + 102 * e) >> 6;
+            uint8_t r  = (uint8_t)(clamp(p1));
```

Comment:
Should we use C++ style casts here and below?

```suggestion
            uint8_t r  = static_cast<uint8_t>(clamp(p1));
```

Reply:
Removed the CUPVA based algorithm source code and related documentation for now.
It will be added in the upcoming releases.